spectator-ext-stackdriver/src/main/java/com/netflix/spectator/stackdriver/StackdriverWriter.java

```diff
@@ -0,0 +1,760 @@
+  }
+
+  /**
+   * Helper function to encode a value into buffer when constructing a UUID.
+   */
+  private void addLong(byte[] buffer, int offset, long value) {
+    for (int i = 0; i < 8; ++i) {
+      buffer[i + offset] = (byte)(value >> (8 * i) & 0xff);
+    }
+  }
+
+  /**
+   * [Re]bind the instanceId used to identify the TimeSeries data coming from
+   * this particular instance.
+   */
+  public void setInstanceId(String id) {
+    log.info("Configuring {}={}", INSTANCE_LABEL, id);
+    instanceId = id;
+  }
+    
+  /**
+   * Only useful for testing.
+   */
+  public String getInstanceId() {
+    return instanceId;
+  }
+
+  /**
+   * Lookup all the pre-existing custom metrics for spinnaker so that we
+   * know if we need to create new metric descriptors or not.
+   */
+  public void initKnownDescriptors()
+      throws HttpResponseException, IOException {
+    ListMetricDescriptorsResponse response =
+      service.projects().metricDescriptors().list(this.projectResourceName)
+      .execute();
+
+    for (MetricDescriptor descriptor : response.getMetricDescriptors()) {
+      if (descriptor.getType().startsWith(customTypePrefix)) {
+        knownDescriptors.put(descriptor.getType(), descriptor);
+      }
+    }
+  }
+
+  /**
+   * Inject a hints about one or more types.
+   *
+   * Hints are only used if and when we need to create a type descriptor for
+   * a type. This happens at most once over the lifetime of a system (not just
+   * a process invocation). Adding hints is a pre-emtive workaround for known
+   * inconsistent uses for given types where an incomplete descriptor may
+   * otherwise be inferred.
+   *
+   * Note that the name of the hint is just the Spectator name, not the
+   * Stackdriver type name. That is, it does not yet contain the customTypePrefix.
+   * This is for [assumed] convienence specifying the hints.
+   */
+  public void addCustomDescriptorHints(List<CustomDescriptorHint> hints) {
+    for (CustomDescriptorHint hint : hints) {
+      addLabelHints(hint.name, hint.labels);
+    }
+  }
+
+  /**
+   * A simpler form of addCustomDescriptorHints.
+   */
+  public void addLabelHints(String name, List<String> labels) {
+   if (!name.startsWith(customTypePrefix)) {
+     name = customTypePrefix + name;
+   }
+   List<String> hints = labelHints.get(name);
+   if (hints == null) {
+     hints = new ArrayList<String>(labels.size());
+     hints.addAll(labels);
+     labelHints.put(name, hints);
+   } else {
+     hints.addAll(labels);
+   }
+  }
+
+  /**
+   * Helper function to fetch an individual descriptor from Stackdriver.
+   * This is used to lookup whether or not a particular type already exists.
+   * Normally we would already know this from initKnownDescriptors, however
+   * there is a race condition on first usage where the descriptor did not
+   * exist on our startup but another process has since created it.
+   * Fetching it will correct for that.
```

Comment:
Same question as above on read volume. I'm wondering if these would be prone to thundering herd problems when spinning up large clusters or just having a lot of metrics that we need to fetch descriptors for. AWS has fairly strict limits and will start throttling, I'm not as familiar with stackdriver, but it is a concern.